src/main/java/org/nypl/findawayclientlibrary/DownloadService.java

```diff
@@ -141,19 +183,253 @@ public void onNext(Integer progress) {
+   * NOTE: Audio files are downloaded and stored under the application's standard internal files directory. This directory is deleted when the application is removed.
+   *
+   * NOTE:  As long as we use the DownloadEngine to download, queue, delete, etc., the Engine will have stored knowledge
+   * of the current state in the findaway internal database.  So, if I use the engine to delete, it will change the status to “NOT_DOWNLOADED”
+   * which I can check for and re-download later.
+   *
+   * NOTE:  Call getDownloadEngine().getProgress(contentId) to get the download progress
+   * of a book as a percentage from 0 to 100.  Or use the download event system.
+   *
+   * TODO:  Test how to request downloads for 2different books (may need to make my own queue).
+   *
+   * @param contentId  the catalog book id
+   * @param contentId  id of the license to play book (what was checked out to patron)
+   * @param chapter  the chapter to start download at
+   * @param part  the more specific chapter part to start download at
+   */
+  public Integer downloadAudio(String contentId, String licenseId, Integer chapter, Integer part) {
+    if (part == null) {
+      // most of the time, we don't need to get granular.  start download at part==0, and auto-proceed from there.
+      part = CHAPTER_PART_DEFAULT;
+    }
+
+    if ((contentId == null) || (licenseId == null) || (chapter == null)) {
+      LogHelper.e(TAG, "DownloadService cannot build download request for (contentId, licenseId, chapter)", contentId, licenseId, chapter);
+      return DOWNLOAD_ERROR;
+    }
+
+    // NOTE:  DownloadType.TO_END gets book from specified chapter to the end of the book.
+    // DownloadType.TO_END_WRAP gets from specified chapter to the end of the book, then wraps around and
+    // gets all the beginning chapters, too.
+    // DownloadType.SINGLE gets just that chapter.
+    // The system does skip any chapters that are already downloaded.  So, if we need to re-download a chapter,
+    // we'd have to delete it first then call download.
+
+    LogHelper.d(TAG, "before building downloadRequest, part=" + part + ", chapter=" + chapter);
+    // NOTE:  If you start with the first chapter in your request the DownloadEngine will skip any chapters that are already downloaded.
+    DownloadRequest downloadRequest = DownloadRequest.builder().contentId(contentId).part(part).chapter(chapter).licenseId(licenseId).type(DownloadType.TO_END_WRAP).build();
+
+    try {
+      this.getDownloadEngine().download(downloadRequest);
+      LogHelper.d(TAG, "after downloadEngine.download \n\n\n");
+    } catch (Exception e) {
+      LogHelper.e(TAG, e, "Error getting download engine: " + e.getMessage());
+      return DOWNLOAD_ERROR;
+    }
+
+    return DOWNLOAD_RUNNING;
+  }
+
+
+  /**
+   * Delete all content previously downloaded for this book.
+   * TODO: untested
+   *
+   * @param contentId  book to delete
+   * @return
+   */
+  public Integer deleteDownload(String contentId) {
```

Comment:
I'm not sure I understand why this method is returning an `Integer`, even if that `Integer` is a pseudo-enum. Most of the possible enum values make no sense here. It would be good to limit the return type so it only allows sensible return values so that the caller doesn't have to anticipate handling things that can't actually happen.

Reply:
Not sure what would be the better return strategy.  I'd like to let the calling code know whether the delete operation succeeded.  This can be achieved by returning a flag, or throwing an exception, so I chose to return a flag.  Once the delete is actually done, the download engine is expected to throw a notifying event, which we then catch in line 588.